File: apple/Sources/SwiftyKit/Session.swift
import CaladonCoreFFI
import CryptoKit
import Foundation

/// Confidential session channel — WMK delivery into the CVM (identity-envelope.md §6).
/// Now a thin wrapper over the shared `caladon-core` Rust trust core (`CaladonCoreFFI`):
/// X25519 ECDH + HKDF-SHA256 session-key derivation and the XChaCha20-Poly1305 sealed
/// envelopes (WMK delivery + live "chat" turns) all run in Rust, byte-identical to
/// `swifty_crypto.session` and the prior CryptoKit/libsodium implementation, so the client
/// and CVM derive the SAME session key and the WMK opens in TEE RAM. The KDF info binds BOTH
/// endpoints' public keys (anti-UKS/MITM) inside the core.
public enum Session {
    public static let label = Data("swifty/session/v1".utf8)
    public static let wmkDeliveryPurpose = "wmk-delivery"
    static let keyLength = 32
    static let nonceLength = 24  // XChaCha20-Poly1305 nonce; the FFI seal returns `nonce ‖ ct`

    public static let chatPurpose = "chat"  // live-turn prompt/response envelopes sealed to SK

    public enum SessionError: Error { case badPublicKeyLength, lowOrderPoint }

    /// Fresh ephemeral X25519 keypair: (privateBytes[32], publicBytes[32]). The client's
    /// ephemeral pub doubles as the attestation challenge (SHA-256(pub)) and the ECDH input.
    /// Stays platform-native (CryptoKit CSPRNG) — only the resulting bytes flow into the core.
    public static func x25519Keypair() -> (privateBytes: Data, publicBytes: Data) {
        let priv = Curve25519.KeyAgreement.PrivateKey()
        return (priv.rawRepresentation, priv.publicKey.rawRepresentation)
    }

    /// X25519 public key (32 bytes) for an ephemeral private scalar.
    public static func x25519Public(privateBytes: Data) throws -> Data {
        do {
            return try CaladonCoreFFI.x25519Public(privateBytes: privateBytes)
        } catch let e as CaladonCoreFFI.SealError {
            throw map(e)
        }
    }

    /// SK = HKDF(X25519(myPrivate, theirPublic), info = label ‖ clientPub ‖ cvmPub).
    /// Both sides pass the SAME clientPub/cvmPub and their own private + the peer's public.
    /// Fails closed on a bad public-key length or a low-order/identity peer point (the core
    /// rejects the all-zero shared secret, matching Python's `cryptography`).
    public static func deriveSessionKey(
        myPrivate: Data, theirPublic: Data, clientPub: Data, cvmPub: Data
    ) throws -> Data {
        do {
            return try CaladonCoreFFI.deriveSessionKey(
                myPrivate: myPrivate, theirPublic: theirPublic, clientPub: clientPub, cvmPub: cvmPub
            )
        } catch let e as CaladonCoreFFI.SealError {
            throw map(e)
        }
    }

    /// Client seals WMK to the session key for delivery (§6). Returns (nonce, ct); the aad
    /// binds account_id ‖ purpose ‖ v inside the core, exactly as `swifty_crypto.seal_wmk`.
    public static func sealWMK(sessionKey: Data, wmk: Data, accountID: String, v: Int = 1) throws
        -> (nonce: [UInt8], ct: [UInt8])
    {
        do {
            let nonceCt = try CaladonCoreFFI.sealWmk(
                sessionKey: sessionKey, wmk: wmk, accountId: accountID, v: Int64(v)
            )
            return splitNonceCt(nonceCt)
        } catch let e as CaladonCoreFFI.SealError {
            throw map(e)
        }
    }

    /// CVM-side open of the sealed WMK; fails closed on any tamper / wrong session. `v` MUST
    /// be the version carried by the RECEIVED envelope (no default — the aad binds `v`, so a
    /// wrong `v` silently fails the open; the caller must source it from the wire envelope,
    /// exactly as Python's `open_wmk` reads `envelope["v"]`).
    public static func openWMK(
        sessionKey: Data, nonce: [UInt8], ct: [UInt8], accountID: String, v: Int
    ) throws -> Data {
        do {
            return try CaladonCoreFFI.openWmk(
                sessionKey: sessionKey, nonceCt: joinNonceCt(nonce, ct), accountId: accountID, v: Int64(v)
            )
        } catch let e as CaladonCoreFFI.SealError {
            throw map(e)
        }
    }

    /// Seal a live-turn payload (prompt or delta) to SK (purpose "chat"). Returns (nonce, ct).
    public static func sealChat(sessionKey: Data, plaintext: Data, accountID: String, v: Int = 1) throws
        -> (nonce: [UInt8], ct: [UInt8])
    {
        do {
            let nonceCt = try CaladonCoreFFI.sealChat(
                sessionKey: sessionKey, plaintext: plaintext, accountId: accountID, v: Int64(v)
            )
            return splitNonceCt(nonceCt)
        } catch let e as CaladonCoreFFI.SealError {
            throw map(e)
        }
    }

    /// Open a sealed live-turn payload (a response delta) under SK. Fails closed on tamper.
    public static func openChat(sessionKey: Data, nonce: [UInt8], ct: [UInt8], accountID: String, v: Int) throws -> Data {
        do {
            return try CaladonCoreFFI.openChat(
                sessionKey: sessionKey, nonceCt: joinNonceCt(nonce, ct), accountId: accountID, v: Int64(v)
            )
        } catch let e as CaladonCoreFFI.SealError {
            throw map(e)
        }
    }

    // MARK: - FFI framing + error mapping

    /// The core returns/consumes `nonce ‖ ct`; SwiftyKit's public surface keeps the (nonce, ct)
    /// tuple (the gateway wire envelope carries the nonce separately).
    private static func splitNonceCt(_ nonceCt: Data) -> (nonce: [UInt8], ct: [UInt8]) {
        (nonce: [UInt8](nonceCt.prefix(nonceLength)), ct: [UInt8](nonceCt.dropFirst(nonceLength)))
    }

    private static func joinNonceCt(_ nonce: [UInt8], _ ct: [UInt8]) -> Data {
        Data(nonce) + Data(ct)
    }

    private static func map(_ e: CaladonCoreFFI.SealError) -> SessionError {
        switch e {
        case .LowOrderPoint: return .lowOrderPoint
        // BadKey / BadNonce / OpenFailed / SealFailed all surface here as a fail-closed
        // key/length rejection (the only non-low-order path the public API can hit pre-open
        // is a bad public-key length).
        default: return .badPublicKeyLength
        }
    }
}

File: gateway/app/session.py
"""Confidential session channel — SERVER side (identity-envelope.md §6).

In `cvm` mode the in-CVM gateway is INSIDE the attested boundary, so it legitimately holds a
per-session working-memory key (WMK) in TEE RAM. It derives the session key
`SK = HKDF(X25519(cvm_session_priv, client_eph_pub), info = label ‖ client_pub ‖ cvm_pub)`,
opens the client-sealed WMK (XChaCha20-Poly1305), uses it for the session, and **never
persists it** (dropped on `end()` / process exit).

Byte-identical to the reference `swifty_crypto.session` / `.envelope` (parity-tested against
their vectors). Established primitives only: X25519 + HKDF-SHA256 via `cryptography`,
XChaCha20-Poly1305 via libsodium (`pynacl`) — `cryptography` has only IETF ChaCha20 (12-byte
nonce), not the 24-byte XChaCha20 the envelope uses.
"""

from __future__ import annotations

import dataclasses
import hashlib
import hmac
import threading

import nacl.bindings as sodium
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey, X25519PublicKey
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

SESSION_LABEL = b"swifty/session/v1"
WMK_DELIVERY_PURPOSE = "wmk-delivery"
CHAT_PURPOSE = "chat"  # live-turn prompt/response envelopes sealed to SK
KEY_LEN = 32
NONCE_LEN = 24  # XChaCha20-Poly1305-IETF


def x25519_keypair() -> tuple[bytes, bytes]:
    """(private_bytes[32], public_bytes[32]) for the CVM's ephemeral session key."""
    priv = X25519PrivateKey.generate()
    return priv.private_bytes_raw(), priv.public_key().public_bytes_raw()


def derive_session_key(my_private: bytes, their_public: bytes, *, client_pub: bytes, cvm_pub: bytes) -> bytes:
    """SK from ECDH + HKDF, binding BOTH endpoints' public keys (anti-UKS/MITM). Matches
    swifty_crypto.session.derive_session_key exactly so client and CVM agree."""
    if len(client_pub) != KEY_LEN or len(cvm_pub) != KEY_LEN:
        raise ValueError("public keys must be 32 bytes")
    if len(my_private) != KEY_LEN or len(their_public) != KEY_LEN:
        raise ValueError("session keys must be 32 bytes")
    shared = X25519PrivateKey.from_private_bytes(my_private).exchange(
        X25519PublicKey.from_public_bytes(their_public)
    )
    info = SESSION_LABEL + client_pub + cvm_pub  # fixed-length pubs -> unambiguous
    return HKDF(algorithm=hashes.SHA256(), length=KEY_LEN, salt=None, info=info).derive(shared)


def _aad(account_id: str, purpose: str, v: int) -> bytes:
    # Matches swifty_crypto.envelope._aad (newline-delimited, SHA-256).
    return hashlib.sha256(f"{account_id}\n{purpose}\n{v}".encode()).digest()


def _open(session_key: bytes, *, nonce: bytes, aad: bytes, ct: bytes, v: int, account_id: str, purpose: str) -> bytes:
    """Open an SK-sealed blob. Fail-closed: re-derives the AAD from the authoritative
    account_id/purpose/v and constant-time-checks it against the wire AAD BEFORE the AEAD
    open (so a blob can't be replayed across accounts/purposes), then XChaCha20 decrypts (the
    tag also covers the AAD). Any tamper raises."""
    expected = _aad(account_id, purpose, v)
    if not hmac.compare_digest(expected, aad):
        raise ValueError("aad does not bind to this account/purpose/v")
    return sodium.crypto_aead_xchacha20poly1305_ietf_decrypt(ct, expected, nonce, session_key)


def _seal(session_key: bytes, plaintext: bytes, *, account_id: str, purpose: str, v: int = 1) -> tuple[bytes, bytes, bytes]:
    """Seal under SK with a fresh CSPRNG nonce; aad binds account_id‖purpose‖v. Returns
    (nonce, aad, ct)."""
    aad = _aad(account_id, purpose, v)
    nonce = sodium.randombytes(NONCE_LEN)
    ct = sodium.crypto_aead_xchacha20poly1305_ietf_encrypt(plaintext, aad, nonce, session_key)
    return nonce, aad, ct


def open_wmk(session_key: bytes, *, nonce: bytes, aad: bytes, ct: bytes, v: int, account_id: str) -> bytes:
    """Open the client-sealed WMK (purpose 'wmk-delivery')."""
    return _open(session_key, nonce=nonce, aad=aad, ct=ct, v=v, account_id=account_id, purpose=WMK_DELIVERY_PURPOSE)


@dataclasses.dataclass
class _Session:
    sk: bytes
    wmk: bytes

    def __repr__(self) -> str:  # never let SK/WMK reach a log/traceback/error-monitor
        return "<_Session sk=REDACTED wmk=REDACTED>"


class SessionManager:
    """Per-process holder of the CVM's session keypair + per-account established sessions
    {SK, WMK}. WMK lives ONLY in memory (TEE RAM analog) — never written to disk; dropped on
    `end()` or process exit (§6). Thread-safe."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._cvm_priv, self._cvm_pub = x25519_keypair()  # ephemeral, per process, WRITE-ONCE
        self._sessions: dict[str, _Session] = {}

    def __repr__(self) -> str:  # redact: no SK/WMK in logs/tracebacks
        return f"<SessionManager sessions={len(self._sessions)} cvm_pub={self._cvm_pub.hex()[:16]}…>"

    @property
    def cvm_pub(self) -> bytes:
        """The CVM session public key the client needs (delivered, attestation-bound, in the
        evidence) to derive SK."""
        return self._cvm_pub

    def establish(self, account_id: str, client_eph_pub: bytes, *, nonce: bytes, aad: bytes, ct: bytes, v: int) -> None:
        """Derive SK from the client's ephemeral pub + open the sealed WMK into TEE RAM."""
        # Snapshot the WRITE-ONCE CVM keypair; safe to use lock-free (never reassigned). If
        # keypair rotation is ever added, move this read inside the lock.
        cvm_priv, cvm_pub = self._cvm_priv, self._cvm_pub
        sk = derive_session_key(cvm_priv, client_eph_pub, client_pub=client_eph_pub, cvm_pub=cvm_pub)
        wmk = open_wmk(sk, nonce=nonce, aad=aad, ct=ct, v=v, account_id=account_id)
        # Crypto runs lock-free (pure, no shared state); only the store is locked. Concurrent
        # establishes for the SAME account are content-idempotent (a tenant's WMK is
        # deterministic from its seed), so a last-write-wins overwrite is harmless.
        with self._lock:
            self._sessions[account_id] = _Session(sk=sk, wmk=wmk)

    def wmk_for(self, account_id: str) -> bytes | None:
        with self._lock:
            s = self._sessions.get(account_id)
            return s.wmk if s else None

    def has_session(self, account_id: str) -> bool:
        with self._lock:
            return account_id in self._sessions

    def open_for(self, account_id: str, *, nonce: bytes, aad: bytes, ct: bytes, v: int, purpose: str = CHAT_PURPOSE) -> bytes:
        """Open an SK-sealed blob (e.g. the live-turn prompt) for an established session.
        Raises KeyError if no session, or on any tamper (fail-closed). The session lookup +
        AEAD open happen UNDER the lock as one step, so SK never leaves the locked region and
        a concurrent end() can't slip between the check and the open (XChaCha20 of a small
        blob is microseconds — holding the lock is negligible)."""
        with self._lock:
            s = self._sessions.get(account_id)
            if s is None:
                raise KeyError("no established session")
            return _open(s.sk, nonce=nonce, aad=aad, ct=ct, v=v, account_id=account_id, purpose=purpose)

    def seal_for(self, account_id: str, plaintext: bytes, *, purpose: str = CHAT_PURPOSE, v: int = 1) -> tuple[bytes, bytes, bytes]:
        """Seal a response delta under the session SK (under the lock; SK never leaves).
        Raises KeyError if no session."""
        with self._lock:
            s = self._sessions.get(account_id)
            if s is None:
                raise KeyError("no established session")
            return _seal(s.sk, plaintext, account_id=account_id, purpose=purpose, v=v)

    def end(self, account_id: str) -> None:
        with self._lock:
            self._sessions.pop(account_id, None)
